packages/nx/src/project-graph/utils/project-configuration-utils.ts

```diff
@@ -31,53 +36,148 @@ export function mergeProjectConfigurationIntoRootMap(
-  if (project.tags && matchingProject.tags) {
+  if (project.tags) {
     updatedProjectConfiguration.tags = Array.from(
-      new Set(matchingProject.tags.concat(project.tags))
+      new Set((matchingProject?.tags ?? []).concat(project.tags))
     );
+
+    project.tags.forEach((tag) => {
+      projectConfigSourceMap?.set(
+        `${project.root}.tags.${tag}`,
+        sourceInformation
+      );
+    });
   }
 
-  if (project.implicitDependencies && matchingProject.implicitDependencies) {
-    updatedProjectConfiguration.implicitDependencies =
-      matchingProject.implicitDependencies.concat(project.implicitDependencies);
+  if (project.implicitDependencies) {
+    updatedProjectConfiguration.implicitDependencies = (
+      matchingProject?.implicitDependencies ?? []
+    ).concat(project.implicitDependencies);
+
+    project.implicitDependencies.forEach((implicitDependency) => {
+      projectConfigSourceMap?.set(
+        `${project.root}.implicitDependencies.${implicitDependency}`,
+        sourceInformation
+      );
+    });
   }
 
-  if (project.generators && matchingProject.generators) {
+  if (project.generators) {
     // Start with generators config in new project.
     updatedProjectConfiguration.generators = { ...project.generators };
 
-    // For each generator that was already defined, shallow merge the options.
-    // Project contains the new info, so it has higher priority.
-    for (const generator in matchingProject.generators) {
-      updatedProjectConfiguration.generators[generator] = {
-        ...matchingProject.generators[generator],
-        ...project.generators[generator],
-      };
+    for (const generator in project.generators) {
+      projectConfigSourceMap?.set(
+        `${project.root}.generators.${generator}`,
+        sourceInformation
+      );
+      for (const property in project.generators[generator]) {
+        projectConfigSourceMap?.set(
+          `${project.root}.generators.${generator}.${property}`,
+          sourceInformation
+        );
+      }
+    }
+
+    if (matchingProject.generators) {
+      // For each generator that was already defined, shallow merge the options.
+      // Project contains the new info, so it has higher priority.
+      for (const generator in matchingProject.generators) {
+        updatedProjectConfiguration.generators[generator] = {
+          ...matchingProject.generators[generator],
+          ...project.generators[generator],
+        };
+
+        for (const property in project.generators[generator]) {
+          projectConfigSourceMap?.set(
+            `${project.root}.generators.${generator}.${property}`,
+            sourceInformation
+          );
+        }
+      }
     }
   }
 
-  if (project.namedInputs && matchingProject.namedInputs) {
+  if (project.namedInputs) {
     updatedProjectConfiguration.namedInputs = {
-      ...matchingProject.namedInputs,
+      ...(matchingProject?.namedInputs ?? {}),
```

Comment:
The guarding here is unnecessary, we know matching project isn't null or undefined because we set it if it wasn't already set, and inside of an object you can spread null / undefined without any issue.
```suggestion
      ...matchingProject.namedInputs,
```